core-bundle/src/Resources/contao/library/Contao/System.php

```diff
@@ -618,41 +624,57 @@ public static function getCountries()
 	 */
 	public static function getLanguages($blnInstalledOnly=false)
 	{
-		$return = array();
-		$languages = array();
-		$arrAux = array();
-		$langsNative = array();
+		$request = self::getContainer()->get('request_stack')->getCurrentRequest();
+		$currentLocale = $request ? $request->getLocale() : 'en';
 
-		static::loadLanguageFile('languages');
-		include __DIR__ . '/../../config/languages.php';
+		$locales = array_values(self::getContainer()->getParameter('contao.locales'));
 
-		foreach ($languages as $strKey=>$strName)
+		if (!$blnInstalledOnly)
 		{
-			$arrAux[$strKey] = isset($GLOBALS['TL_LANG']['LNG'][$strKey]) ? Utf8::toAscii($GLOBALS['TL_LANG']['LNG'][$strKey]) : $strName;
+			$locales = array_values(array_unique(array_merge($locales, \ResourceBundle::getLocales(''))));
 		}
 
-		asort($arrAux);
-
-		$arrBackendLanguages = self::getContainer()->getParameter('contao.locales');
+		static::loadLanguageFile('languages');
 
-		foreach (array_keys($arrAux) as $strKey)
-		{
-			if ($blnInstalledOnly && !\in_array($strKey, $arrBackendLanguages))
+		$return = array_map(
+			static function ($locale) use ($currentLocale)
 			{
-				continue;
-			}
+				$label = $GLOBALS['TL_LANG']['LNG'][$locale] ?? \Locale::getDisplayName($locale, $currentLocale) ?: $locale;
+				$labelNative = \Locale::getDisplayName($locale, $locale);
 
-			$return[$strKey] = $GLOBALS['TL_LANG']['LNG'][$strKey] ?? $languages[$strKey];
+				if (!$labelNative || $label === $labelNative)
+				{
+					return $label;
+				}
 
-			if (isset($langsNative[$strKey]) && $langsNative[$strKey] != $return[$strKey])
-			{
-				$return[$strKey] .= ' - ' . $langsNative[$strKey];
-			}
-		}
+				return $label . ' - ' . $labelNative;
+			},
+			array_combine($locales, $locales)
+		);
+
+		asort($return);
 
 		// HOOK: add custom logic
 		if (isset($GLOBALS['TL_HOOKS']['getLanguages']) && \is_array($GLOBALS['TL_HOOKS']['getLanguages']))
 		{
+			trigger_deprecation('contao/core-bundle', '4.12', 'Using the "getLanguages" hook has been deprecated and will no longer work in Contao 5.0. Extend the contao.locales parameter instead.');
```

Comment:
With `contao.locales` you can only change the list of languages available in various places globally. However with the hook you can change this dynamically, depending on the context. Do we really want to remove this possibility?

Reply:
The backend locales did not use this method/hook anymore in many places. With the current state of this pull request, you can configure `contao.backend_locales` and `contao.intl.locales` and additionaly decorate the locales service.